### PR TITLE
feat(dataloaders): add argument for custom data split indexes

### PR DESCRIPTION
**Pull request by ori-kron-wis**
a branch to handle several data splitting issues

### CODE REVIEW

File: src/scvi/dataloaders/_data_splitting.py [hunk 1]
```diff
@@ -22,7 +22,11 @@
 
 
 def validate_data_split(
-    n_samples: int, train_size: float, validation_size: Optional[float] = None
+    n_samples: int,
+    train_size: float,
+    validation_size: Optional[float] = None,
+    use_external_indexing: Optional[bool] = False,
+    external_indexing: Optional[list] = None,
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:29]: I like the use of `external_indexing` to contain all data split indices instead of having separate arguments for training and validation. To make it more concise, we could remove `use_external_indexing` and assume that `external_indexing` will be used when it is not `None`

File: src/scvi/dataloaders/_data_splitting.py [hunk 2]
```diff
@@ -34,27 +38,77 @@ def validate_data_split(
         Size of train set. Need to be: 0 < train_size <= 1.
     validation_size
         Size of validation set. Need to be 0 <= validation_size < 1
+    use_external_indexing
+        Whether to use external supported indexing. This bypass any other flag or input parameter
+        that was before
+    external_indexing
+        A list of np.arrays that is always in the order of [[train_idx],[valid_idx],[test_idx]].
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:45]:
```suggestion
        A list of data split indexes in the order of training, validation, and test sets.
```
nit: Since our docs will render type annotations directly, we can remove type references in argument docstrings for less redundancy.

Comment (martinkim0) [reply src/scvi/dataloaders/_data_splitting.py:45]: We should also mention that validation and test are not required

File: src/scvi/dataloaders/_data_splitting.py [hunk 3]
```diff
@@ -34,27 +38,77 @@ def validate_data_split(
...
     """
-    if train_size > 1.0 or train_size <= 0.0:
-        raise ValueError("Invalid train_size. Must be: 0 < train_size <= 1")
+    if use_external_indexing:
+        if external_indexing is None:
+            raise ValueError("External indexing list is missing")
+
+        if type(external_indexing) is not list:
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:51]: nit: For our code style it's preferable to use `not isinstance(external_indexing, list))`

Comment (martinkim0) [reply src/scvi/dataloaders/_data_splitting.py:51]: Applicable to other lines in this function as well.

File: src/scvi/dataloaders/_data_splitting.py [hunk 4]
```diff
@@ -34,27 +38,77 @@ def validate_data_split(
...
+            raise ValueError("External indexing is not of list type")
+
+        # validate the structure of it
+        if type(external_indexing[0]) is not np.ndarray:
+            raise ValueError("The first element of the list of lists is not a np.array")
+
+        if external_indexing[1] is None:
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:58]: This will fail if `len(external_indexing) < 2)`, same for the `external_indexing[2] is None` check below.

File: src/scvi/dataloaders/_data_splitting.py [hunk 5]
```diff
@@ -34,27 +38,77 @@ def validate_data_split(
...
+    if use_external_indexing:
+        if external_indexing is None:
+            raise ValueError("External indexing list is missing")
+
+        if type(external_indexing) is not list:
+            raise ValueError("External indexing is not of list type")
+
+        # validate the structure of it
+        if type(external_indexing[0]) is not np.ndarray:
+            raise ValueError("The first element of the list of lists is not a np.array")
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:56]: For the sake of not repeating these checks explicitly for each element, you can do something like:
```
if not all(isinstance(elem, np.ndarray) for elem in external_indexing)):
   # raise error here
```
No need to specify which element is not the correct type - I think the users can figure that out.

File: src/scvi/dataloaders/_data_splitting.py [hunk 6]
```diff
@@ -34,27 +38,77 @@ def validate_data_split(
...
+
+        # check for total number of indexes (overlapping or missing)
+        if (
+            len(external_indexing[0]) + len(external_indexing[1]) + len(external_indexing[2])
+        ) > n_samples:
+            raise ValueError("There are overlapping indexing please fix")
+        if (
+            len(external_indexing[0]) + len(external_indexing[1]) + len(external_indexing[2])
+        ) < n_samples:
+            raise ValueError("There are missing indexing please fix or remove those lines")
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:79]: I think this scenario OK to have - some users may wish to leave out some observations. We can still raise a warning here though.

File: src/scvi/dataloaders/_data_splitting.py (hunk 1 above)
Comment (martinkim0) [reply src/scvi/dataloaders/_data_splitting.py:29]: We can also make the typing more explicit here with `Optional[list[npt.ArrayLike]]`

File: src/scvi/dataloaders/_data_splitting.py [hunk 7]
```diff
@@ -118,24 +181,38 @@ def __init__(
...
+        self.use_external_indexing = use_external_indexing
+        self.external_indexing = external_indexing
 
         self.n_train, self.n_val = validate_data_split(
-            self.adata_manager.adata.n_obs, self.train_size, self.validation_size
+            self.adata_manager.adata.n_obs,
+            self.train_size,
+            self.validation_size,
+            self.use_external_indexing,
+            self.external_indexing,
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:192]: Since `validate_data_split` is now trying to handle two sets of arguments (one for `train_size` and the other for `external_indexing`), maybe we can refactor this out into two functions instead. What do you think?

File: tests/dataloaders/test_datasplitter.py [hunk 8]
```diff
@@ -40,6 +40,38 @@ def test_datasplitter_shuffle():
     )
 
 
+def test_datasplitter_external():
```
Comment (martinkim0) [inline tests/dataloaders/test_datasplitter.py:43]: Ideally we should also be testing for error cases (overlapping indexes, missing input arguments, etc). You can do this with `with pytest.raises`.

File: src/scvi/dataloaders/_data_splitting.py [hunk 9]
```diff
@@ -34,27 +38,77 @@ def validate_data_split(
...
+        if type(external_indexing[1]) is not np.ndarray:
+            raise ValueError("The second element of the list of lists is not a np.array")
+        if type(external_indexing[2]) is not np.ndarray:
+            raise ValueError("The third element of the list of lists is not a np.array")
+
+        # check for total number of indexes (overlapping or missing)
+        if (
+            len(external_indexing[0]) + len(external_indexing[1]) + len(external_indexing[2])
+        ) > n_samples:
+            raise ValueError("There are overlapping indexing please fix")
```
Comment (martinkim0) [inline src/scvi/dataloaders/_data_splitting.py:75]: An edge case for this check is if any of the indexes contains duplicate elements, e.g., the train split having `[1, 2, 3, 1]`.

Comment (martinkim0) [reply src/scvi/dataloaders/_data_splitting.py:75]: One way to address this would be to convert each element in the input list to a numpy set. This will also make the `np.intersect1d` checks below faster.

**Review by martinkim0 — DISMISSED** [review]
Great start! One major comment regarding refactoring the logic here into two separate functions, and some minor comments about code style, error checks, and testing.

Could you also add a release note under `CHANGELOG.md` in the same format as the other entries? This can go under the `1.2.0` release.

Also, if you have time, feel free to use the newer Python 3.10+ type annotations in `src/scvi/dataloaders/_data_splitting.py` by adding a `from __future__ import annotations` at the top of the file.

**Comment by martinkim0** [comment]
I'm also going to go ahead and change the title of the PR to match our [commit convention](https://docs.scvi-tools.org/en/latest/developer/maintenance.html#conventional-commits).

File: src/scvi/dataloaders/_data_splitting.py (hunk 1 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:29]: Make sense

File: src/scvi/dataloaders/_data_splitting.py (hunk 2 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:45]: gotcha. done.

File: src/scvi/dataloaders/_data_splitting.py (hunk 3 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:51]: OK I see.

File: src/scvi/dataloaders/_data_splitting.py (hunk 5 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:56]: Sure, I actually added here handling for the case where the 2nd or 3rd arrays are even not given to us, and imputed that . But made it now more compact.

File: src/scvi/dataloaders/_data_splitting.py (hunk 4 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:58]: it is now different

File: src/scvi/dataloaders/_data_splitting.py (hunk 6 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:79]: alright

File: src/scvi/dataloaders/_data_splitting.py (hunk 9 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:75]: I would keep only the intersection part and add np.unique(s)

I didn't think about duplicates.
The problem is not only here, the user can insert duplicate indexes per set (not overlapping) and its train/test/valid will hold duplicate rows. I don't think it makes sense. but maybe he will want to up/down sample for some reason.

I added to raise a warning for duplicates per set

File: src/scvi/dataloaders/_data_splitting.py (hunk 7 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:192]: Can you elaborate? Wouldn't this be much more challenging to implement and manage?

Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py]: change to sets in the beginning of validation and use that further on code.
divide only the validation part to 2 cases
We want to implement it in the semi-supervised class as well (divide to labeld/unlabled)

File: src/scvi/dataloaders/_data_splitting.py [hunk 10]
```diff
@@ -52,13 +52,74 @@ def validate_data_split(
...
+    if len(external_indexing_unique[2]) < len(external_indexing[2]):
+        raise Warning("There are duplicate indexing in test set")
+
+    # check for total number of indexes (overlapping or missing)
+    if (
+        len(external_indexing_unique[0])
+        + len(external_indexing_unique[1])
+        + len(external_indexing_unique[2])
+    ) < n_samples:
+        raise Warning("There are missing indexing please fix or remove those lines")
```
Comment (canergen) [inline src/scvi/dataloaders/_data_splitting.py:108]: indexing = indices

File: src/scvi/dataloaders/_data_splitting.py [hunk 11]
```diff
@@ -83,6 +144,9 @@ class DataSplitter(pl.LightningDataModule):
     pin_memory
         Whether to copy tensors into device-pinned memory before returning them. Passed
         into :class:`~scvi.data.AnnDataLoader`.
+    external_indexing
+        A list of data split indexes in the order of training, validation, and test sets.
+        Validation and test set and not required and can be left empty.
```
Comment (canergen) [inline src/scvi/dataloaders/_data_splitting.py:149]: and not = are not.
Left empty = setting it to None or not using it?

File: src/scvi/dataloaders/_data_splitting.py [hunk 12]
```diff
@@ -260,53 +345,78 @@ def __init__(
...
-            n_labeled_train, n_labeled_val = validate_data_split(
-                n_labeled_idx, self.train_size, self.validation_size
-            )
-
-            labeled_permutation = self._labeled_indices
-            if self.shuffle_set_split:
-                rs = np.random.RandomState(seed=settings.seed)
-                labeled_permutation = rs.choice(
-                    self._labeled_indices, len(self._labeled_indices), replace=False
+            # Need to separate tp the external and non-external cases of the labeled indices
```
Comment (canergen) [inline src/scvi/dataloaders/_data_splitting.py:356]: what's tp?

File: src/scvi/dataloaders/_data_splitting.py [hunk 13]
```diff
@@ -260,53 +345,78 @@ def __init__(
...
-                    len(self._unlabeled_indices),
-                    replace=False,
+            # Need to separate tp the external and non-external cases of the unlabeled indices
+            if self.external_indexing is not None:
+                # we need to intersect the external indexing given with the labeled indices
+                unlabeled_idx_train, unlabeled_idx_val, unlabeled_idx_test = (
+                    np.intersect1d(self.external_indexing[n], self._unlabeled_indices)
+                    for n in range(3)
+                )
+                n_labeled_train, n_labeled_val = validate_data_split_with_external_indexing(
```
Comment (canergen) [inline src/scvi/dataloaders/_data_splitting.py:397]: Should be n_unlabeled_train?

File: src/scvi/external/gimvi/_model.py [hunk 14]
```diff
@@ -202,6 +203,9 @@ def train(
         plan_kwargs
             Keyword args for model-specific Pytorch Lightning task. Keyword arguments passed
             to `train()` will overwrite values present in `plan_kwargs`, when appropriate.
+        external_indexing
+            A list of data split indexes in the order of training, validation, and test sets.
+            Validation and test set and not required and can be left empty.
```
Comment (canergen) [inline src/scvi/external/gimvi/_model.py:208]: and =are
indexes = indices

File: tests/dataloaders/test_datasplitter.py [hunk 15]
```diff
@@ -40,6 +40,79 @@ def test_datasplitter_shuffle():
...
+    adata = scvi.data.synthetic_iid()
```
Comment (canergen) [inline tests/dataloaders/test_datasplitter.py:44]: We should test here also all modules with specific exceptions. I.e. pyro, contrastive, scanvi, scvi. Those tests should go in the respective module tests. The error with n_unlabelled above should throw an error in the scANVI test but not in the tests here.

**Review by canergen — COMMENTED** [review]
Please add the requested tests and fix the typos in the documentation and explain how to set indices to empty in the docstring. Afterwards fine and can be merged.

File: src/scvi/dataloaders/_data_splitting.py (hunk 10 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:108]: done

File: src/scvi/dataloaders/_data_splitting.py (hunk 11 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:149]: Both will work actually.

File: src/scvi/dataloaders/_data_splitting.py (hunk 12 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:356]: type - need to be to

File: src/scvi/external/gimvi/_model.py (hunk 14 above)
Comment (ori-kron-wis) [reply src/scvi/external/gimvi/_model.py:208]: fixed

File: src/scvi/dataloaders/_data_splitting.py (hunk 13 above)
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_data_splitting.py:397]: yes missed that

File: tests/dataloaders/test_datasplitter.py (hunk 15 above)
Comment (ori-kron-wis) [reply tests/dataloaders/test_datasplitter.py:44]: Yes you are correct.
I will add those and it will be a good option to see how to run a different module.

**Review by canergen — APPROVED** [review]

**Comment by canergen** [comment]
Hi @martinkim0 and @ori-kron-wis does it have to be merged again(?) after adding on-merge: backport to 1.2.X?
